### PR TITLE
Implement renderYAML

### PR DESCRIPTION
**Pull request by matheus23**
Hi! This is a first attempt at https://discourse.dhall-lang.org/t/expense-proposal-pure-dhall-function-to-render-yaml/92.

I've wanted to get started with Dhall some time ago and this seemed like a nice first project.

As this was my first Dhall project, there might be some things that I've gotten completely wrong, that's why I'm reporting my progress early. This is only a prototype and I'd love some review.

There are some things this is doing differently than `libyaml`. It seems libyaml removes duplicate fields in objects, my prototype doesn't do that. Also, Double literals are rendered as `1.0` instead of `1`.

I haven't tested many edge cases, as I don't know how to get some real-world test data, so I'd love support in that direction!

### CODE REVIEW

**Comment by matheus23** [comment]
Here is a test file I used to test this with:
```dhall
let JSON = ./Prelude/JSON/package.dhall

let renderYAML = ./Prelude/JSON/renderYAML

let renderedYAML =
      renderYAML
        ( JSON.array
            [ JSON.bool True
            , JSON.string "Hello"
            , JSON.object
                [ { mapKey = "foo", mapValue = JSON.null }
                , { mapKey = "bar", mapValue = JSON.number 1.0 }
                , { mapKey = "object"
                  , mapValue =
                      JSON.object
                        [ { mapKey = "empty"
                          , mapValue =
                              JSON.object
                                ( [] : List
                                         { mapKey : Text, mapValue : JSON.Type }
                                )
                          }
                        , { mapKey = "empty"
                          , mapValue = JSON.array ([] : List JSON.Type)
                          }
                        ]
                  }
                ]
            , JSON.array [ JSON.bool True, JSON.bool False ]
            ]
        )

in  ''
    # Is this YAML yet?
    ${renderedYAML}
    ''
```
```sh
$ dhall --file test.dhall
''
Is this YAML yet?
- true
- "Hello"
- foo: null
  bar: 1.0
  object:
    empty: {}
    empty: []
- - true
  - false
''
```

**Comment by matheus23** [comment]
I have been taking a closer look at YAML escaping rules.

I wanted to write up a small post about what I found out, but it turns out, [its really complicated, and someone else already wrote a lot about it](https://stackoverflow.com/questions/3790454/how-do-i-break-a-string-over-multiple-lines/21699210#21699210).

In the worst case (multiline strings with trailing whitespace), `libyaml` simply resorts to using single-line double-quotation as an escaping mechanism:

``` yaml
key: value
? "key with multiline: string \"\nand - special characters "
: null
```
(despite github's yaml syntax highlighting messing this up, this is `libyaml`-generated)

For everything in between, there are many ways to handle it: Various multiline formats that preserve or don't preserve whitespace and single-quoted strings with an exclamation mark.

I'm unsure how to proceed.
Using `Text/show` it would be possible to generate (probably) always valid, but very ugly YAML.

If I had some primitive for checking for the existence of certain properties of a given `Text`, like leading/trailing whitespace or the `": "` or `"- "` characters for example, then I could only conditionally YAML-escape keys and values.

What to do now depends heavily on the usecase:
* Do such strings that need to be escaped even come up in the real world? Does it make sense to handle these cases if these edge cases practically would never come up?
* Or do these edge cases come up but nobody looks at the generated YAML files anyway, so whether they're pretty doesn't matter?
* Or do these edge cases ocassionally come up and we should try to have some minimum effort (double-quoted escaping) for supporting them?

**Comment by sjakobi** [comment]
I think we should just go with `Text/show` for rendering strings. Without the ability of  checking for certain substrings, I think there's nothing else that we can do and that would still work in the general case.

> If I had some primitive for checking for the existence of certain properties of a given `Text`, like leading/trailing whitespace or the `": "` or `"- "` characters for example, then I could only conditionally YAML-escape keys and values.

AFAIK there are no plans for adding such primitives since they would be difficult to standardize.

File: Prelude/JSON/renderYAML [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+
+let List/null = ../List/null
+
+let List/map = ../List/map
+
+let List/concat = ../List/concat
+
+let List/concatMap = ../List/concatMap
+
+let uncons =
```
Comment (Gabriel439) [inline Prelude/JSON/renderYAML]: Also, if we introduce a `NonEmpty` then the type of this function could change to:

```dhall
uncons : ∀(a : Type) → List a → Optional (NonEmpty a)
```

File: Prelude/JSON/renderYAML [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+let blockToText
+    : Block → Text
+    = λ(block : Block) → Text/concatSep "\n" (blockLines block)
+
+let render
+    : JSON → Text
+    =   λ(json : JSON)
+      → let ObjectField = { mapKey : Text, mapValue : Block }
+        
+        in  blockToText
```
Comment (Gabriel439) [inline Prelude/JSON/renderYAML]: Perhaps add a trailing newline at the outermost level

File: Prelude/JSON/renderYAML [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+              [ head.tail, List/concatMap Block Text blockLines tail ]
+        }
+
+let manyBlocks
+    : Block → List Block → Block
+    =   λ(ifEmpty : Block)
+      → λ(blocks : List Block)
+      → uncons Block blocks Block ifEmpty concatBlocks
+
+let singleline
```
Comment (Gabriel439) [inline Prelude/JSON/renderYAML]: Minor suggestion: Perhaps capitalize this as `singleLine` for consistency with the other identifiers

File: Prelude/JSON/renderYAML (hunk 2 above)
Comment (matheus23) [reply Prelude/JSON/renderYAML]: Good idea!

File: Prelude/JSON/renderYAML [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+        λ(headIndent : Text)
+      → λ(tailIndent : Text)
+      → λ(block : Block)
+      → { head = headIndent ++ block.head
+        , tail = List/map Text Text (λ(t : Text) → tailIndent ++ t) block.tail
+        }
+
+let blockLines
+    : Block → List Text
+    = λ(block : Block) → List/concat Text [ [ block.head ], block.tail ]
```
[the remark this answers is not in the document]
Comment (matheus23) [reply Prelude/JSON/renderYAML]: Ah yes. Some feedback: It was weird to be both looking at the wiki cheatsheet page and the Prelude directory for functions that do what I want. The cheatsheet seems to contain some functions from the Prelude, but not all, which was confusing. (Maybe it is just outdated? In some places it still uses the old `[ x ] : Optional x` syntax).

**Comment by matheus23** [comment]
I factored out a `NonEmpty` type with `uncons`, `nonEmptyToList` and `concatNonEmpty` functions. It might be useful to add those to the Prelude at some point, right?

Also, I'm having a hard time understanding what's wrong in the CI. I added integrity checks to the imports and added `renderYAML` to the `package.dhall`. What else do I have to do?

**Comment by sjakobi** [comment]
> I factored out a `NonEmpty` type with `uncons`, `nonEmptyToList` and `concatNonEmpty` functions. It might be useful to add those to the Prelude at some point, right?

I would welcome such an addition to the `Prelude`! :)



> Also, I'm having a hard time understanding what's wrong in the CI. I added integrity checks to the imports and added `renderYAML` to the `package.dhall`. What else do I have to do?

I think you need freeze the local imports too. Try `dhall freeze --cache --all`.

And `Prelude/package.dhall` needs to be updated with the new hash for the `JSON` `package.dhall`.

**Comment by sjakobi** [comment]
> Some feedback: It was weird to be both looking at the wiki cheatsheet page and the Prelude directory for functions that do what I want. The cheatsheet seems to contain some functions from the Prelude, but not all, which was confusing. (Maybe it is just outdated? In some places it still uses the old `[ x ] : Optional x` syntax).

The cheatsheet documents only the built-in functions – I think it would get too large if we'd document the entire Prelude there. We do need better documentation for the Prelude too though! (See https://github.com/dhall-lang/dhall-lang/issues/760.)

We should definitely fix the outdated syntax though!

**Comment by Gabriel439** [comment]
@matheus23: Yeah, there is currently no documentation for the Prelude (other than the documentation within the files).  Dhall does not currently have a documentation generator analogous to Haskell's `haddock`

If you have Nix installed you can satisfy CI by running the `./scripts/lint-prelude.sh` script.

Even if you don't have Nix installed, you can still see what you need to do to fix the files here:

https://github.com/dhall-lang/dhall-lang/blob/da02ab2faf81b3e627f01cc5f4bb37403a15ab0e/release.nix#L78-L87

I also just fixed the outdated `Optional` syntax in the Cheatsheet

File: Prelude/JSON/renderYAML [hunk 5]
```diff
@@ -0,0 +1,218 @@
...
+
+   The generated YAML text will only contain escaped object keys and
+   string values and might therefore not be very human readable.
+
+   However, it is useful for debugging `JSON` values or for tests.
+   For anything more sophisticated you should use `dhall-to-json` or
+   `dhall-to-yaml`.
+
+```
+let JSON = ./package.dhall
```
Comment (Gabriel439) [inline Prelude/JSON/renderYAML]: Now that you have an `assert` below you no longer need the same example here

File: Prelude/JSON/renderYAML [hunk 6]
```diff
@@ -0,0 +1,218 @@
...
+                              elements
+                          )
+                  , bool =
+                      λ(x : Bool) → singleLine (if x then "true" else "false")
+                  , null = singleLine "null"
+                  }
+              )
+
+let example0 =
+      let JSON =
```
Comment (Gabriel439) [inline Prelude/JSON/renderYAML]: Perhaps move this `JSON` binding to the top of the file, for two reasons:

* For consistency in having all imports near the top of the file
* So that we could potentially reuse the same `JSON` binding for multiple `example`s

Alternatively, we could provide a JSON record containing just the JSON construction fields (i.e. a `./build.dhall` which is essentially this record), and then modify the `./package.dhall` to re-export that record, like this:

```dhall
-- ./package.dhall
let build = ./build.dhall

in    { Type =
            ./Type sha256:5adb234f5868a5b0eddeb034d690aaba8cb94ea20d0d557003e90334fff6be3e
          ? ./Type
      , Tagged =
            ./Tagged sha256:21feca7d2b23f210d0696131d792e18a7d24fdcc85d41a49ba85b98670eba194
          ? ./Tagged
      , Nesting =
            ./Nesting sha256:6284802edd41d5d725aa1ec7687e614e21ad1be7e14dd10996bfa9625105c335
          ? ./Nesting
      , …
      }
    ∧ build
```

That way we could reuse `./build.dhall` for other examples within this directory without incurring an import cycle.

Comment (matheus23) [reply Prelude/JSON/renderYAML]: Hm. I nested this let statement as I was already shadowing a `let JSON = ./Type` from above.

I could merge these two let bindings by including the JSON Type. Should a `./build` file export the `Prelude/JSON/Type` as well?

Comment (Gabriel439) [reply Prelude/JSON/renderYAML]: Yeah, I think it makes sense to include the `Type` in the `./build` record, too

**Comment by matheus23** [comment]
I now introduced a `Prelude/JSON/build` file that includes all types and helper functions for creating JSON values.

I changed `omitNullFields` and `JSON.render` so that they now import `./build` so it is consistent between renderYAML and the other modules (and extracted the example from JSON.render out to an assertion, while I was at it).

**Comment by sjakobi** [comment]
I like the structure that the `build` file offers, but I don't really like the name: It looks just like the other exports like `bool` or `null` – although AFAICT it's not intended for public consumption!

Can we rename it to something like `internal.dhall` (with the extension!) to make it clearer that it's not just another function, and users can simply ignore it?

A comment at the top of the file should explain why the file exists.

**Comment by Gabriel439** [comment]
@sjakobi: Note that it's not truly internal because users can still import it.  Maybe `build` is not the right name, but I think it should still have a descriptive name rather than `internal`

**Comment by Gabriel439** [comment]
@sjakobi: `core.dhall` seems fine to me

**Comment by matheus23** [comment]
I agree about `./build` actually. It's also not consistent with `List/build` or `Optional/build` etc.

I also like `core.dhall` most. `constructor.dhall` wouldn't cover `Type` being in this module.

File: Prelude/JSON/core.dhall [hunk 7]
```diff
@@ -0,0 +1,34 @@
+{ Type =
```
Comment (sjakobi) [inline Prelude/JSON/core.dhall]: Could you add a comment explaining the purpose of this record?

Comment (matheus23) [reply Prelude/JSON/core.dhall]: Oh yeah, I missed that. On a second though though, should I? It is basically mostly a re-export of other values. In haskell these modules are also often not documented and `package.dhall`s don't seem to be documented either.

Oh well, I guess it wouldn't hurt.

Comment (sjakobi) [reply Prelude/JSON/core.dhall]: Well I guess the question would be: Why can't the users of `core.dhall` not simply use `package.dhall`?!

File: Prelude/JSON/core.dhall [hunk 8]
```diff
@@ -0,0 +1,44 @@
+{- A record of functions useful for constructing `JSON` values.
```
Comment (sjakobi) [inline Prelude/JSON/core.dhall]: This line reads as if Prelude users ought to use this record.

Comment (Gabriel439) [reply Prelude/JSON/core.dhall]: Actually, maybe they ought to!  It's sort of a historical accident that these `core` utilities are in the same record as other utilites (like `./Nesting`).  If we weren't worried about backwards compatibility I would have reorganized the `./JSON` subdirectory a bit

Comment (sjakobi) [reply Prelude/JSON/core.dhall]: Right! I had noticed the inconsistency in this directory too. How would you organize it?

Comment (matheus23) [reply Prelude/JSON/core.dhall]: I was hoping the sentence after this one would clarify this.

Maybe use `used` instead of `useful`?

I like to have a sentence about what something is in that something's documentation, instead of documentation saying you shouldn't use this. (Documentation saying 'don't use this' doesn't seem like documentation for me).

I know that this module shouldn't be used by Prelude users, but I don't think documentation should be for prelude users only. Contributors might also wonder what the reason for this file is.

**Comment by matheus23** [comment]
I'm having trouble making the CI happy again :/ Locally, I'm running a bash script that I extracted out of the release.nix file to attempt to do something similar to CI:
```bash
#!/bin/bash
for FILE in $(find "Prelude" -type f ! -name README.md); do
    echo "fixing file: $FILE"
    dhall lint --inplace "$FILE"
    dhall format --inplace "$FILE"
    XDG_CACHE_HOME=$(mktemp -d) dhall freeze --all --cache --inplace "$FILE"
done
```

I see there are more checks than running `dhall freeze` on the project, so I've been trying to run `release.nix` locally. Is `nix-build release.nix` the right command for this? I'm hoping I'll have enough space left on my disk and still waiting for that command to finish.

**Comment by sjakobi** [comment]
I can't help much with the Nix tooling here.

The error at https://hydra.dhall-lang.org/build/43325/nixlog/3 is

```
diff --recursive ./tests.actual/type-inference/success/preludeB.dhall ./tests.expected/type-inference/success/preludeB.dhall
171a172,186
>     , renderYAML :
>           ∀ ( json
>             :   ∀(JSON : Type)
>               → ∀ ( json
>                   : { array : List JSON → JSON
>                     , bool : Bool → JSON
>                     , null : JSON
>                     , number : Double → JSON
>                     , object : List { mapKey : Text, mapValue : JSON } → JSON
>                     , string : Text → JSON
>                     }
>                   )
>               → JSON
>             )
>         → Text
builder for '/nix/store/0p6pm3hiycx9lac72r124qkjzi4zy294-test-files-lint.drv' failed with exit code 1
```

So you need to update the `preludeB.dhall` golden file with the new inferred type of the Prelude's `package.dhall`.

File: Prelude/JSON/core.dhall (hunk 8 above)
Comment (sjakobi) [reply Prelude/JSON/core.dhall]: I guess my intuition is that the documentation of the Prelude should be optimized for Prelude _users_. So I think a good first sentence would be 

```suggestion
{- Functions for constructing `JSON` values for internal usage
```

The "internal" should signal to users that this file is probably not interesting to them, so they can quickly move on.

But I realize that this is premature optimization. Sorry for being so nitpicky and thanks for having addressed my concerns so far! :)

**Comment by Gabriel439** [comment]
@matheus23: If you already have Nix installed then we have scripts for this purpose underneath the `./scripts` directory.  Specifically, the script you want is `./scripts/lint-prelude.sh`

Also, you don't have to build everything from scratch if you add our shared Nix cache by following the instructions here: https://github.com/dhall-lang/dhall-haskell#nix

**Comment by Gabriel439** [comment]
@matheus23: Sorry, the correct script is actually `./scripts/generate-test-files.sh`, although it looks like you already fixed it.

Either way this looks ready to merge.  I added you as a collaborator so that you can merge this yourself

**Comment by Gabriel439** [comment]
@matheus23: Thank you for doing this! 🙂
